### ISSUE TITLE
"Error: Error encountered resolving symbol values statically." when console.log called from a forRoot static method"

### ISSUE THREAD

**Issue by Yermo** [angular/angular#17663]
I originally posted this to the ionic issue tracker here: https://github.com/ionic-team/ionic/issues/12113

They suggest that this is probably an Angular issue so I am cross posting here.

## I'm submitting a ...

<pre><code>
[ ] Regression (behavior that used to work and stopped working in a new release)
[x ] Bug report 
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

$ ionic cordova build android --prod 
generates the following error:

```
Error: Error encountered resolving symbol values statically. Calling function 'TestModule', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol AppModule
```

## Expected behavior

I expected the app to build.

## Minimal reproduction of the problem with instructions

ionic start TestBuild tabs

Create directory src/modules.
Create a module, TestModule in src/modules/test.module.ts with contents:

```
import { 
  NgModule, 
  ModuleWithProviders
} from '@angular/core';

export interface ITestConfig {
  value: string;
}

export class TestConfig implements ITestConfig {
  value: string;
}

@NgModule({

  declarations: [
  ],

  imports: [
  ],

  entryComponents: [
  ],

  providers: [
  ],

  exports: [
  ]

})

export class TestModule {

  // -----------------------------------------------------------------------------

  /** 
  * test set configuration
  */

  static forRoot( config: ITestConfig ): ModuleWithProviders {

    console.log( "This is a test" );

    return {
      ngModule: TestModule,
      providers: [

        { provide: TestConfig, useValue: config }

      ]
    };
  }

} // end of TestModule

```
Include module test module in src/app/app.module.ts

```
import { TestModule, TestConfig } from '../modules/test.module';

let config: TestConfig = { 
  value: 'test'
} 

....

  imports: [
    BrowserModule,
    IonicModule.forRoot(MyApp),
    TestModule.forRoot( config )
  ], 
```

Run the build and notice the error.

Now remove the "console.log()" call and re-run the build.

Note that it builds.

Why would a call to console.log there cause the build to fail with this error?

## Please tell us about your environment

<pre><code>
    "@angular/common": "4.1.3",
    "@angular/compiler": "4.1.3",
    "@angular/compiler-cli": "4.1.3",
    "@angular/core": "4.1.3",
    "@angular/forms": "4.1.3",
    "@angular/http": "4.1.3",

yml@cylon:~/Projects/TestBuild$ ionic info

global packages:

    @ionic/cli-utils : 1.4.0
    Cordova CLI      : 6.4.0 
    Ionic CLI        : 3.4.0

local packages:

    @ionic/app-scripts              : 1.3.7
    @ionic/cli-plugin-cordova       : 1.4.0
    @ionic/cli-plugin-ionic-angular : 1.3.1
    Cordova Platforms               : android 6.0.0
    Ionic Framework                 : ionic-angular 3.4.2

System:

    Node       : v7.10.0
    OS         : Linux 4.4
    Xcode      : not installed
    ios-deploy : not installed
    ios-sim    : not installed
    npm        : 4.2.0 

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX 
- Platform: 
node v7.10.0
Ubuntu 16.04.2 LTS
Others:

cordova version 6.4.0

</code></pre>

**Comment by Toxicable** [comment]
The error states: `Error encountered resolving symbol values statically. Calling function 'TestModule', function calls are not supported.`
and in your code you have a function call, highlighted here: 
```
static forRoot( config: ITestConfig ): ModuleWithProviders {

    console.log( "This is a test" ); //This is a function call

    return {
      ngModule: TestModule,
      providers: [
        { provide: TestConfig, useValue: config }
      ]
    };
  }
```

**Comment by Yermo** [comment]
Thank you.

It would seem to me the error message should mention the function call console.log() instead of "TestModule", which is class. That's what confused me.

**Comment by chuckjaz** [comment]
@Yermo Unfortunately, we don't know that you are calling `console.log` at that point. All we know is that you are calling function off of `TestModule` that cannot be evaluated because it didn't qualify as a macro function at the time it was collected.

**Comment by Yermo** [comment]
Thank you for the reply. I appreciate it. As is evident, I'm relatively new to the angular universe. I tend to use console.log liberally to help me figure out how all the pieces fit together.

**Comment by mat3e** [comment]
@Yermo does your code work after deleting a `console.log` call?

**Comment by Yermo** [comment]
@mat3e yes. Works like a champ. (I must say I'm incredibly impressed with Angular.)

**Comment by shairez** [comment]
BTW, that also happens if I declare a variable instead of a function call

It might be useful to change the error message to explain that anything inside the forRoot function that isn't exported could cause the error...

**Comment by MurhafSousli** [comment]
I have the same issue without having `console.log()` in `forRoot()`
```
export class NgProgressModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: NgProgressModule,
      providers: [NgProgress],
    };
  }
}
```
here is the the [module](https://github.com/MurhafSousli/ngx-progressbar/blob/master/lib/core/src/ng-progress.module.ts#L19)

**Comment by chuckjaz** [comment]
@shairez The error message is currently targeted at callers of the method not the author of the method. 99% of the time this message is displayed it is because someone called a method in the decorator that we don't support. Including how to "fix" the method that they called but shouldn't have is not really all that helpful either.

What do you think about, for example, a comment ahead of the method, such as one that contains "@macro" or "@metadata" that would then be a trigger to the collector to include more information about why the method was not collected.

**Comment by shairez** [comment]
@chuckjaz  So you mean - the author of the method will need to add that comment? 

**Just to clarify my last comment -**
I was the same person who also declared and used the method.

I needed to guess and work hard until I connected the dots that this error message was about a variable I referred to from the exported method (it was a factory method as part of some 3rd party ngModule forRoot method ). 

I don't know if it's possible from the compiler point of view, but if we can somehow change the error message to give more info that will be a huge time saver in finding these mistakes.
